### ISSUE TITLE
Remove tokio-sync dependency in tower-limit

### ISSUE THREAD

**Issue by marcelbuesing** [tower-rs/tower#384]
I noticed when looking at tonic that it still has some `futures-preview` dependencies and it turned out to be due to `tower-limit` depending on `tokio-sync = "0.2.0-alpha.6"`. My initial idea for a merge request was just use the `sync` feature of the tokio crate instead but it turned out that the module `semaphore` is private in `tokio 0.2`. 

```
error[E0603]: module `semaphore` is private
  --> tower-limit/src/concurrency/future.rs:11:18
   |
11 | use tokio::sync::semaphore::Semaphore;
```

Apparently the module was made private in [this commit](https://github.com/tokio-rs/tokio/commit/0d38936b35779b604770120da2e98560bbb6241f).

Any ideas in regards how to get rid of tokio-sync = "0.2.0-alpha.6 ?

**Comment by marcelbuesing** [comment]
https://github.com/tokio-rs/tokio/issues/1825

**Comment by LucioFranco** [comment]
At this point, I am contemplating vendoring the sempahore implementation within tower-limit since last I checked with carl he wanted to do more work on it before we made it public. This should allow us to remove the futures-preview crates.

**Comment by seanmonstar** [comment]
Closed via #388
